app/src/main/java/is/xyz/mpv/MPVView.kt

```diff
@@ -23,15 +24,71 @@ internal class MPVView(context: Context, attrs: AttributeSet) : GLSurfaceView(co
         observeProperties()
     }
 
+
     fun initOptions() {
-        MPVLib.setOptionString("hwdec", "mediacodec")
-        MPVLib.setOptionString("vo", "opengl-cb")
+        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this.context)
+
+        // initial options
+        data class Property(val preference_name: String, val mpv_option: String)
+
+        // vo
+        var vo = "opengl-cb"
+
+        val vos = arrayOf(
+                Property("video_upscale","scale"),
+                Property("video_downscale","dscale"),
+                Property("video_scale_param1", "scale-param1"),
+                Property("video_scale_param2", "scale-param2")
+        )
+
+        for ((preference_name, mpv_option) in vos) {
+            var preference = sharedPreferences.getString(preference_name, "")
+            if (!preference.isNullOrBlank())
+                vo += ":$mpv_option=$preference"
+        }
+
+        // hwdec
+        val hardware_decoding = sharedPreferences.getBoolean("hardware_decoding", false)
+        val hwdec = if (hardware_decoding) "mediacodec" else "no"
+
+        // ao
         // set optimal buffer size and sample rate for opensles, to get better audio playback
         val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
         val framesPerBuffer = am.getProperty(AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER)
         val sampleRate = am.getProperty(AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE)
         Log.v(TAG, "Device reports optimal frames per buffer $framesPerBuffer sample rate $sampleRate")
-        MPVLib.setOptionString("ao", "opensles:frames-per-buffer=$framesPerBuffer:sample-rate=$sampleRate")
+
+        var ao = "opensles:frames-per-buffer=$framesPerBuffer:sample-rate=$sampleRate"
+
+        // overrides
+
+        var opengl_custom = sharedPreferences.getString("opengl_custom", "")
```

Comment:
What's the purpose of allowing overriding of vo?
Anything except `opengl-cb` will simply not work.